src/main/java/com/google/devtools/build/lib/analysis/constraints/IncompatibleDeterminingHelper.java

```diff
@@ -54,141 +53,161 @@
-                    .build();
-            ConfiguredTargetValue ctv =
-                (ConfiguredTargetValue) env.getValue(constraintDep.getConfiguredTargetKey());
-            if (ctv == null) {
-              return IncompatibleTargetCreationResult.create(true, null);
-            }
-            constraintProvidersBuilder.add(ctv.getConfiguredTarget());
-          }
-          ImmutableList<TransitiveInfoCollection> constraintProviders =
-              constraintProvidersBuilder.build();
-
-          // Find the constraints that don't satisfy the target platform.
-          ImmutableList<ConstraintValueInfo> invalidConstraintValues =
-              PlatformProviderUtils.constraintValues(constraintProviders).stream()
-                  .filter(cv -> !platformInfo.constraints().hasConstraintValue(cv))
-                  .collect(ImmutableList.toImmutableList());
-
-          if (!invalidConstraintValues.isEmpty()) {
-            return IncompatibleTargetCreationResult.create(
-                false,
-                createIncompatibleRuleConfiguredTarget(
-                    target,
-                    configuration,
-                    configConditions,
-                    IncompatiblePlatformProvider.incompatibleDueToConstraints(
-                        platformLabel, invalidConstraintValues),
-                    rule.getRuleClass(),
-                    transitivePackagesForPackageRootResolution));
-          }
-        }
-      }
+    // Find the constraints that don't satisfy the target platform.
+    ImmutableList<ConstraintValueInfo> invalidConstraintValues =
+        constraintKeys.stream()
+            .map(key -> (ConfiguredTargetValue) constraintValues.get(key))
+            .map(ctv -> PlatformProviderUtils.constraintValue(ctv.getConfiguredTarget()))
+            .filter(cv -> cv != null && !platformInfo.constraints().hasConstraintValue(cv))
+            .collect(toImmutableList());
+    if (invalidConstraintValues.isEmpty()) {
+      return null;
     }
-    return IncompatibleTargetCreationResult.create(false, null);
+
+    return createIncompatibleRuleConfiguredTarget(
+        target,
+        configuration,
+        configConditions,
+        IncompatiblePlatformProvider.incompatibleDueToConstraints(
+            platformInfo.label(), invalidConstraintValues),
+        rule.getRuleClass(),
+        transitivePackagesForPackageRootResolution);
   }
 
+  /**
+   * Creates an incompatible target if it is "indirectly incompatible".
+   *
+   * <p>In other words, this function checks if a target is incompatible because of one of its
+   * dependencies. If a dependency is incompatible, then this target is also incompatible.
+   *
+   * <p>This function returns either null or an incompatible configured target. If null is returned,
+   * the target is not indirectly incompatible and analysis of this target should proceed.
+   */
   @Nullable
   public static RuleConfiguredTargetValue createIndirectlyIncompatibleTarget(
```

Comment:
For simplicity, this should also return an Optional if the previous method does.

Reply:
Done